lib/src/r_tree/quickselect.dart

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) 2021 Ilya Zverev, (c) 2018 Vladimir Agafonkin.
+// Port of https://github.com/mourner/quickselect.
+// Use of this code is governed by an ISC license, see the LICENSE file.
+
+import 'dart:math' as math;
+
+// sort an array so that items come in groups of n unsorted items, with groups sorted between each other;
+// combines selection algorithm with binary divide & conquer approach
+multiSelect<E>(
+    List<E> arr, int left, int right, int n, num Function(E) getter) {
+  final stack = [left, right];
+  final compare = (E a, E b) => getter(a).compareTo(getter(b));
+  while (stack.isNotEmpty) {
```

Comment:
We probably want to check if the stack has at least 2 items, otherwise calling `removeLast` twice on a stack with 1 item can yield errors. 

This is an unlikely scenario given that we seem to always add an even number of items, though I think it is worth the effort to be cautious.

Reply:
In the case of an error in the algorithm I'd rather throw an error than silently prevent multiSelect from running if it gets into a state where stack has 1 item in it. As you said, we start with 2 items and always remove 2, we also only add 4 at a time.